From 7b013295bd8cb84d7b75be37e39ba9e719233618 Mon Sep 17 00:00:00 2001
From: Karthik Kambatla <kasha@cloudera.com>
Date: Tue, 9 Aug 2016 13:42:25 -0700
Subject: [PATCH 1705/2848] HADOOP-13299. JMXJsonServlet is vulnerable to
 TRACE. (Haibo Chen via kasha)

(cherry picked from commit 85422bb7c5d3e70a49f620ba1c8800e0ba4b64f2)

Change-Id: I784d68096c36ffa5db77182df5809de9fcdebf64
---
 .../java/org/apache/hadoop/jmx/JMXJsonServlet.java |    9 +++++++++
 .../org/apache/hadoop/jmx/TestJMXJsonServlet.java  |   15 +++++++++++++++
 2 files changed, 24 insertions(+)

diff --git a/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/jmx/JMXJsonServlet.java b/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/jmx/JMXJsonServlet.java
index ce0e7f1..f486538 100644
--- a/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/jmx/JMXJsonServlet.java
+++ b/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/jmx/JMXJsonServlet.java
@@ -147,6 +147,15 @@ protected boolean isInstrumentationAccessAllowed(HttpServletRequest request,
   }
 
   /**
+   * Disable TRACE method to avoid TRACE vulnerability.
+   */
+  @Override
+  protected void doTrace(HttpServletRequest req, HttpServletResponse resp)
+      throws ServletException, IOException {
+    resp.sendError(HttpServletResponse.SC_METHOD_NOT_ALLOWED);
+  }
+
+  /**
    * Process a GET request for the specified resource.
    * 
    * @param request
diff --git a/hadoop-common-project/hadoop-common/src/test/java/org/apache/hadoop/jmx/TestJMXJsonServlet.java b/hadoop-common-project/hadoop-common/src/test/java/org/apache/hadoop/jmx/TestJMXJsonServlet.java
index 978e907..d994e94 100644
--- a/hadoop-common-project/hadoop-common/src/test/java/org/apache/hadoop/jmx/TestJMXJsonServlet.java
+++ b/hadoop-common-project/hadoop-common/src/test/java/org/apache/hadoop/jmx/TestJMXJsonServlet.java
@@ -30,6 +30,10 @@
 import org.junit.BeforeClass;
 import org.junit.Test;
 
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+import java.net.HttpURLConnection;
+
 public class TestJMXJsonServlet extends HttpServerFunctionalTest {
   private   static final Log LOG = LogFactory.getLog(TestJMXJsonServlet.class);
   private static HttpServer2 server;
@@ -103,4 +107,15 @@ public static void assertReFind(String re, String value) {
     assertReFind("\\}\\);$", result);
 
   }
+
+  @Test
+  public void testTraceRequest() throws IOException {
+    URL url = new URL(baseUrl, "/jmx");
+    HttpURLConnection conn = (HttpURLConnection) url.openConnection();
+    conn.setRequestMethod("TRACE");
+
+    assertEquals("Unexpected response code",
+        HttpServletResponse.SC_METHOD_NOT_ALLOWED, conn.getResponseCode());
+  }
+
 }
-- 
1.7.9.5

